### DISCUSSION TITLE
Trouble with--perhaps unique--dataset design

### DISCUSSION THREAD

**Discussion by bjfranks** [pyg-team/pytorch_geometric#3403] [Q&A] [answered]
Hey,

I am trying to setup my own dataset, where each data point is a pair of graphs as well as a label. This dataset would be somewhat similar to GEDDataset, the difference being I dont have access to a label for each possible pair. Because of this I do not think I can take the same approach. I looked around and decided using HeteroData instead with basically one key for each graph might be better.

One data point looks something like
```data = HeteroData()
 data['G1'].edge_index = graph1.edge_index
 data['G1'].edge_attr = graph1.persistence
 data['G1'].num_nodes = graph1.num_nodes
 data['G2'].edge_index = graph2.edge_index
 data['G2'].edge_attr = graph2.persistence
 data['G2'].num_nodes = graph2.num_nodes
 data.y = c;
```
My first question, is this a reasonable setup? I believe I can then access the relevant graph when I am loading the batches.

With this current setup I am running into a problem though

> D:\mlvis\ML\venv\lib\site-packages\torch_geometric\data\storage.py:264: UserWarning: Unable to accurately infer 'num_nodes' from the attribute set 'set()'. Please explicitly set 'num_nodes' as an attribute of 'data[G]' to suppress this warning
  " to suppress this warning")
D:\mlvis\ML\venv\lib\site-packages\torch_geometric\data\storage.py:264: UserWarning: Unable to accurately infer 'num_nodes' from the attribute set 'set()'. Please explicitly set 'num_nodes' as an attribute of 'data[1]' to suppress this warning
  " to suppress this warning")
Traceback (most recent call last):
  File "datasets.py", line 58, in <module>
    for batch in loader:
  File "D:\mlvis\ML\venv\lib\site-packages\torch\utils\data\dataloader.py", line 517, in __next__
    data = self._next_data()
  File "D:\mlvis\ML\venv\lib\site-packages\torch\utils\data\dataloader.py", line 557, in _next_data
    data = self._dataset_fetcher.fetch(index)  # may raise StopIteration
  File "D:\mlvis\ML\venv\lib\site-packages\torch\utils\data\_utils\fetch.py", line 47, in fetch
    return self.collate_fn(data)
  File "D:\mlvis\ML\venv\lib\site-packages\torch_geometric\loader\dataloader.py", line 39, in __call__
    return self.collate(batch)
  File "D:\mlvis\ML\venv\lib\site-packages\torch_geometric\loader\dataloader.py", line 20, in collate
    self.exclude_keys)
  File "D:\mlvis\ML\venv\lib\site-packages\torch_geometric\data\batch.py", line 75, in from_data_list
    exclude_keys=exclude_keys,
  File "D:\mlvis\ML\venv\lib\site-packages\torch_geometric\data\collate.py", line 86, in collate
    increment)
  File "D:\mlvis\ML\venv\lib\site-packages\torch_geometric\data\collate.py", line 156, in _collate
    incs = get_incs(key, values, data_list, stores)
  File "D:\mlvis\ML\venv\lib\site-packages\torch_geometric\data\collate.py", line 209, in get_incs
    for value, data, store in zip(values, data_list, stores)
  File "D:\mlvis\ML\venv\lib\site-packages\torch_geometric\data\collate.py", line 209, in <listcomp>
    for value, data, store in zip(values, data_list, stores)
  File "D:\mlvis\ML\venv\lib\site-packages\torch_geometric\data\hetero_data.py", line 268, in __inc__
    return torch.tensor(store.size()).view(2, 1)
RuntimeError: Could not infer dtype of NoneType

The first two issues of course relate to the fact that I am storing two seperate graphs, however they appear to only be warnings. I am also confused why one data point looks like this
```
HeteroData(
  y=0.8116569584235549,
  (G, to, 1)={
    edge_index=[2, 30],
    edge_attr=[30],
    num_nodes=16
  },
  (G, to, 2)={
    edge_index=[2, 22],
    edge_attr=[22],
    num_nodes=12
  }
)
```
The (G, to, 1) part confuses me slightly.

Any ideas wht I am doing or understanding wrong?

**Comment by wsad1** [comment, marked as answer]
I think a good way to store your data is as a pair of graphs in a `Data` object as described [here](https://pytorch-geometric.readthedocs.io/en/latest/notes/batching.html#pairs-of-graphs), for pairs without a label you could store some dummy value.

A `HeretogenousData` object is built to store a graph with different node and edge types, as described [here](https://pytorch-geometric.readthedocs.io/en/latest/notes/heterogeneous.html#creating-heterogeneous-graphs).  So in the code you pasted this line ` data['G1'].edge_index = graph1.edge_index` indicates that an edge is being created from node type `G1` to node type `1` (1 as you didn't specify the other node type).

**Reply by bjfranks** [reply]
Thanks, I was looking for the "proper" way to define this type of object and didnt realize how easy it could be.
